### PR TITLE
CORE-17433 Sandbox Cache thread safety

### PR DESCRIPTION
**Pull request by simon-johnson-r3**
From logs locally and in our resiliency tests it's clear sandbox closing and therefore bundle uninstallation can be performed in multiple threads depending on whether `purgeExpiryQueue` is called from the cache's executor thread via the removal listener, or as part of some other synchronous cache operation which happens in the thread of the client calling it.

For example, with some extra logging we see:
```
[SecManagerForkJoinPool-worker-0] INFO  net.corda.sandboxgroupcontext.service.impl.SandboxGroupContextCacheImpl {...} - @@@ removal listener SecManagerForkJoinPool-worker-0
[SecManagerForkJoinPool-worker-0] INFO  net.corda.sandboxgroupcontext.service.impl.SandboxGroupContextCacheImpl {...} - Closing flow sandbox for O=Bob, L=London, C=GB
[SecManagerForkJoinPool-worker-0] INFO  net.corda.sandboxgroupcontext.service.impl.SandboxGroupContextServiceImpl {...} - @@@ closing SecManagerForkJoinPool-worker-0
```
and also (in the same flow worker):
```
[Thread-37] INFO  net.corda.sandboxgroupcontext.service.impl.SandboxGroupContextCacheImpl {...} - Closing flow sandbox for O=Bob, L=London, C=GB
[Thread-37] INFO  net.corda.sandboxgroupcontext.service.impl.SandboxGroupContextServiceImpl {...} - @@@ closing Thread-37
```

The parallelism in `SecManagerForkJoinPool` is configured to the number of processors. This means we wont just see the members of `SandboxGroupContextCacheImpl` manipulated across 2 threads, we will see it across many and it could happen concurrently in periods of significant flow activity.

This change makes the sandbox cache single threaded, meaning we only ever see manipulation of `SandboxGroupContextCacheImpl` on the same thread - confirmed by examining the logs.

The author of Caffeine states here that using the same thread for the executor is quite acceptable, the Cache itself does little work there: https://stackoverflow.com/questions/68101454/caffeine-combining-both-scheduler-and-executor-service

### CODE REVIEW

**Review by dickon — APPROVED** [review]
LGTM

I agree with Simon that we don't know if this will work but it looks worth a try.

**Review by driessamyn — APPROVED** [review]
LGTM

File: libs/cache/cache-caffeine/src/main/kotlin/net/corda/cache/caffeine/CacheFactory.kt [hunk 1]
```diff
@@ -10,15 +10,25 @@ import com.github.benmanes.caffeine.cache.LoadingCache
...
     /**
-     * Build a Caffeine Loading Cache
+     * Build a Caffeine Cache which executes all actions including callbacks synchronously in the same thread
+     * the calls to the cache are made. The primary use case for this builder is adding non thread safe listeners
+     * to the cache.
+     *
+     * @param name Cache name, used for logging and metrics.
+     * @param caffeine Caffeine cache builder.
+     */
+    fun <K, V> buildNonAsync(name: String, caffeine: Caffeine<in K, in V>): Cache<K, V>
```
Comment (jujoramos) [inline libs/cache/cache-caffeine/src/main/kotlin/net/corda/cache/caffeine/CacheFactory.kt:28]: Nitpick: I'd rather call it `buildSync` 🤷‍♂️.

**Review by jujoramos — APPROVED** [review]
LGTM (I'm running some load testing with the changes from this `PR` to check whether the original issue reproduces or not, will update the `PR` once the test is done).

File: libs/cache/cache-caffeine/src/main/kotlin/net/corda/cache/caffeine/CacheFactory.kt (hunk 1 above)
Comment (simon-johnson-r3) [reply libs/cache/cache-caffeine/src/main/kotlin/net/corda/cache/caffeine/CacheFactory.kt:28]: I mulled over that originally and decided as the default behaviour of Caffeine was async, this made it clearer what the difference was. I don't mind changing it though, I'm not tied to either name.

Comment (jujoramos) [reply libs/cache/cache-caffeine/src/main/kotlin/net/corda/cache/caffeine/CacheFactory.kt:28]: Naa, not need to change it, the argument is sound to me.

**Comment by simon-johnson-r3** [comment]
> LGTM (I'm running some load testing with the changes from this `PR` to check whether the original issue reproduces or not, will update the `PR` once the test is done).

Excellent. Thanks!

**Comment by jujoramos** [comment]
I've ran the test and the changes don't seem to have fixed the original issue (it's worth merging this `PR` in my opinion anyways):
```
{
    "instant":
    {
        "epochSecond": 1697114790,
        "nanoOfSecond": 499000000
    },
    "thread": "single-threaded-scheduled-executor-1",
    "level": "WARN",
    "loggerName": "net.corda.sandboxgroupcontext.service.impl.CurrentSandboxGroupContextImpl",
    "message": "No current sandbox group context to remove for thread [single-threaded-scheduled-executor-1]",
    "thrown":
    {
        "message": "No current sandbox group context to remove for thread",
        "name": "java.lang.IllegalStateException",
        "extendedStackTrace": "java.lang.IllegalStateException: No current sandbox group context to remove for thread
	at net.corda.sandboxgroupcontext.service.impl.CurrentSandboxGroupContextImpl.remove(CurrentSandboxGroupContextImpl.kt:60)
	at net.corda.flow.fiber.FlowFiberImpl.removeCurrentSandboxGroupContext(FlowFiberImpl.kt:236)
	at net.corda.flow.fiber.FlowFiberImpl.suspend(FlowFiberImpl.kt:128)
	at net.corda.flow.fiber.FlowFiberImpl.failTopLevelSubFlow(FlowFiberImpl.kt:191)
	at net.corda.flow.fiber.FlowFiberImpl.runFlow(FlowFiberImpl.kt:109)
	at net.corda.flow.fiber.FlowFiberImpl.run(FlowFiberImpl.kt:61)
	at net.corda.flow.fiber.FlowFiberImpl.run(FlowFiberImpl.kt:20)
	at co.paralleluniverse.fibers.Fiber.run1(Fiber.java:1167)
	at co.paralleluniverse.fibers.Fiber.exec(Fiber.java:863)
	at co.paralleluniverse.fibers.RunnableFiberTask.doExec(RunnableFiberTask.java:100)
	at co.paralleluniverse.fibers.RunnableFiberTask.run(RunnableFiberTask.java:91)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at co.paralleluniverse.concurrent.util.ScheduledSingleThreadExecutor$ScheduledFutureTask.access$001(ScheduledSingleThreadExecutor.java:160)
	at co.paralleluniverse.concurrent.util.ScheduledSingleThreadExecutor$ScheduledFutureTask.run(ScheduledSingleThreadExecutor.java:272)
	at co.paralleluniverse.concurrent.util.ScheduledSingleThreadExecutor.work(ScheduledSingleThreadExecutor.java:119)
	at java.base/java.lang.Thread.run(Thread.java:833)
"
    },
    "endOfBatch": false,
    "loggerFqcn": "org.apache.logging.slf4j.Log4jLogger",
    "threadId": 28,
    "threadPriority": 5
}

{
    "instant":
    {
        "epochSecond": 1697114295,
        "nanoOfSecond": 113000000
    },
    "thread": "single-threaded-scheduled-executor-1",
    "level": "WARN",
    "loggerName": "net.corda.flow.fiber.FlowFiberImpl",
    "message": "Flow failed",
    "thrown":
    {
        "message": "class net.corda.ledger.utxo.flow.impl.transaction.UtxoSignedTransactionImpl cannot be cast to class net.corda.ledger.utxo.flow.impl.transaction.UtxoSignedLedgerTransaction (net.corda.ledger.utxo.flow.impl.transaction.UtxoSignedTransactionImpl and net.corda.ledger.utxo.flow.impl.transaction.UtxoSignedLedgerTransaction are in unnamed module of loader org.apache.felix.framework.BundleWiringImpl$BundleClassLoader @4c41a177)",
        "name": "java.lang.ClassCastException",
        "extendedStackTrace": "
java.lang.ClassCastException: class net.corda.ledger.utxo.flow.impl.transaction.UtxoSignedTransactionImpl cannot be cast to class net.corda.ledger.utxo.flow.impl.transaction.UtxoSignedLedgerTransaction (net.corda.ledger.utxo.flow.impl.transaction.UtxoSignedTransactionImpl and net.corda.ledger.utxo.flow.impl.transaction.UtxoSignedLedgerTransaction are in unnamed module of loader org.apache.felix.framework.BundleWiringImpl$BundleClassLoader @4c41a177)
	at net.corda.ledger.utxo.flow.impl.flows.backchain.TransactionBackchainVerifierImpl.verify(TransactionBackchainVerifierImpl.kt:86)
	at net.corda.ledger.utxo.flow.impl.flows.backchain.v1.TransactionBackchainResolutionFlowV1.call(TransactionBackchainResolutionFlowV1.kt:65)
	at net.corda.ledger.utxo.flow.impl.flows.backchain.v1.TransactionBackchainResolutionFlowV1.call(TransactionBackchainResolutionFlowV1.kt:24)
	at net.corda.flow.application.services.impl.FlowEngineImpl.subFlow(FlowEngineImpl.kt:60)
	at net.corda.flow.application.versioning.impl.ReceiveVersioningFlow.call(ReceiveVersioningFlow.kt:119)
	at net.corda.flow.application.services.impl.FlowEngineImpl.subFlow(FlowEngineImpl.kt:60)
	at net.corda.flow.application.versioning.impl.VersioningServiceImpl.versionedSubFlow(VersioningServiceImpl.kt:38)
	at net.corda.ledger.utxo.flow.impl.flows.backchain.TransactionBackchainResolutionFlow.call(TransactionBackchainResolutionFlow.kt:26)
	at net.corda.ledger.utxo.flow.impl.flows.backchain.TransactionBackchainResolutionFlow.call(TransactionBackchainResolutionFlow.kt:15)
	at net.corda.flow.application.services.impl.FlowEngineImpl.subFlow(FlowEngineImpl.kt:60)
	at net.corda.ledger.utxo.flow.impl.flows.finality.v1.UtxoReceiveFinalityFlowV1.receiveTransactionAndBackchain(UtxoReceiveFinalityFlowV1.kt:121)
	at net.corda.ledger.utxo.flow.impl.flows.finality.v1.UtxoReceiveFinalityFlowV1.call(UtxoReceiveFinalityFlowV1.kt:59)
	at net.corda.ledger.utxo.flow.impl.flows.finality.v1.UtxoReceiveFinalityFlowV1.call(UtxoReceiveFinalityFlowV1.kt:36)
	at net.corda.flow.application.services.impl.FlowEngineImpl.subFlow(FlowEngineImpl.kt:60)
"
    },
    "endOfBatch": false,
    "loggerFqcn": "org.apache.logging.slf4j.Log4jLogger",
    "threadId": 28,
    "threadPriority": 5
}
```

**Comment by simon-johnson-r3** [comment]
> I've ran the test and the changes don't seem to have fixed the original issue (it's worth merging this `PR` in my opinion anyways):

Actually I think it has fixed a bug, but it turns out the ticket describes two different ones. It's eradicated the problem with unloading the bundle - which is what it was targeting. What it hasn't done is fixed the first exception though, which likely has a different cause.

The logs in Loki now differ from the ones in the ticket:

New (from Loki):
```
--
	"thrown": {
		"message": "class net.corda.ledger.utxo.flow.impl.transaction.UtxoSignedTransactionImpl cannot be cast to class net.corda.ledger.utxo.flow.impl.transaction.UtxoSignedLedgerTransaction (net.corda.ledger.utxo.flow.impl.transaction.UtxoSignedTransactionImpl and net.corda.ledger.utxo.flow.impl.transaction.UtxoSignedLedgerTransaction are in unnamed module of loader org.apache.felix.framework.BundleWiringImpl$BundleClassLoader @4c41a177)",
		"name": "java.lang.ClassCastException",
		"extendedStackTrace": "java.lang.ClassCastException: class net.corda.ledger.utxo.flow.impl.transaction.UtxoSignedTransactionImpl cannot be cast to class net.corda.ledger.utxo.flow.impl.transaction.UtxoSignedLedgerTransaction (net.corda.ledger.utxo.flow.impl.transaction.UtxoSignedTransactionImpl and net.corda.ledger.utxo.flow.impl.transaction.UtxoSignedLedgerTransaction are in unnamed module of loader org.apache.felix.framework.BundleWiringImpl$BundleClassLoader @4c41a177)\n\tat net.corda.ledger.utxo.flow.impl.flows.backchain.TransactionBackchainVerifierImpl.verify(TransactionBackchainVerifierImpl.kt:86)\n\tat net.corda.ledger.utxo.flow.impl.flows.backchain.v1.TransactionBackchainResolutionFlowV1.call(TransactionBackchainResolutionFlowV1.kt:65)\n\tat net.corda.ledger.utxo.flow.impl.flows.backchain.v1.TransactionBackchainResolutionFlowV1.call(TransactionBackchainResolutionFlowV1.kt:24)\n\tat net.corda.flow.application.services.impl.FlowEngineImpl.subFlow(FlowEngineImpl.kt:60)\n\tat net.corda.flow.application.versioning.impl.ReceiveVersioningFlow.call(ReceiveVersioningFlow.kt:119)\n\tat net.corda.flow.application.services.impl.FlowEngineImpl.subFlow(FlowEngineImpl.kt:60)\n\tat net.corda.flow.application.versioning.impl.VersioningServiceImpl.versionedSubFlow(VersioningServiceImpl.kt:38)\n\tat net.corda.ledger.utxo.flow.impl.flows.backchain.TransactionBackchainResolutionFlow.call(TransactionBackchainResolutionFlow.kt:26)\n\tat net.corda.ledger.utxo.flow.impl.flows.backchain.TransactionBackchainResolutionFlow.call(TransactionBackchainResolutionFlow.kt:15)\n\tat net.corda.flow.application.services.impl.FlowEngineImpl.subFlow(FlowEngineImpl.kt:60)\n\tat net.corda.ledger.utxo.flow.impl.flows.finality.v1.UtxoReceiveFinalityFlowV1.receiveTransactionAndBackchain(UtxoReceiveFinalityFlowV1.kt:121)\n\tat net.corda.ledger.utxo.flow.impl.flows.finality.v1.UtxoReceiveFinalityFlowV1.call(UtxoReceiveFinalityFlowV1.kt:59)\n\tat net.corda.ledger.utxo.flow.impl.flows.finality.v1.UtxoReceiveFinalityFlowV1.call(UtxoReceiveFinalityFlowV1.kt:36)\n\tat net.corda.flow.application.services.impl.FlowEngineImpl.subFlow(FlowEngineImpl.kt:60)\n\tat net.corda.flow.application.versioning.impl.ReceiveVersioningFlow.call(ReceiveVersioningFlow.kt:119)\n\tat net.corda.flow.application.services.impl.FlowEngineImpl.subFlow(FlowEngineImpl.kt:60)\n\tat net.corda.flow.application.versioning.impl.VersioningServiceImpl.versionedSubFlow(VersioningServiceImpl.kt:38)\n\tat net.corda.ledger.utxo.flow.impl.flows.finality.UtxoReceiveFinalityFlow.call(UtxoReceiveFinalityFlow.kt:27)\n\tat net.corda.ledger.utxo.flow.impl.flows.finality.UtxoReceiveFinalityFlow.call(UtxoReceiveFinalityFlow.kt:16)\n\tat net.corda.flow.application.services.impl.FlowEngineImpl.subFlow(FlowEngineImpl.kt:60)\n\tat net.corda.ledger.utxo.flow.impl.UtxoLedgerServiceImpl.receiveFinality(UtxoLedgerServiceImpl.kt:149)\n\tat com.r3.dc.workflows.token.transfer.TransferTokenFlowResponder.call(TransferTokenFlowResponder.kt:24)\n\tat com.r3.dc.workflows.token.transfer.TransferTokenFlowResponder.call(TransferTokenFlowResponder.kt:13)\n\tat net.corda.flow.application.services.impl.FlowEngineImpl.subFlow(FlowEngineImpl.kt:60)\n\tat com.r3.dc.workflows.token.transfer.TransferTokenAppFlows$Responder.call(TransferTokenAppFlows.kt:144)\n\tat net.corda.flow.fiber.InitiatedFlow.invoke(InitiatedFlow.kt:14)\n\tat net.corda.flow.fiber.FlowFiberImpl.runFlow(FlowFiberImpl.kt:94)\n\tat net.corda.flow.fiber.FlowFiberImpl.run(FlowFiberImpl.kt:61)\n\tat net.corda.flow.fiber.FlowFiberImpl.run(FlowFiberImpl.kt:20)\n\tat co.paralleluniverse.fibers.Fiber.run1(Fiber.java:1167)\n\tat co.paralleluniverse.fibers.Fiber.exec(Fiber.java:863)\n\tat co.paralleluniverse.fibers.RunnableFiberTask.doExec(RunnableFiberTask.java:100)\n\tat co.paralleluniverse.fibers.RunnableFiberTask.run(RunnableFiberTask.java:91)\n\tat java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)\n\tat java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)\n\tat co.paralleluniverse.concurrent.util.ScheduledSingleThreadExecutor$ScheduledFutureTask.access$001(ScheduledSingleThreadExecutor.java:160)\n\tat co.paralleluniverse.concurrent.util.ScheduledSingleThreadExecutor$ScheduledFutureTask.run(ScheduledSingleThreadExecutor.java:272)\n\tat co.paralleluniverse.concurrent.util.ScheduledSingleThreadExecutor.work(ScheduledSingleThreadExecutor.java:119)\n\tat java.base/java.lang.Thread.run(Thread.java:833)\n"
	}
--
	"loggerName": "net.corda.ledger.utxo.flow.impl.flows.backchain.v1.TransactionBackchainResolutionFlowV1",
	"message": "Backchain resolution of [SHA-256D:6F674524488656045C86246550A7DD6B3742F65A16A2838B79A682CE8D746529, SHA-256D:F556198019EB22160672AADC186DB95A7195FCBDD45D9670763E3284E197995A] - Failed due to a transaction in its backchain failing verification",
--
	"loggerName": "net.corda.sandboxgroupcontext.service.impl.CurrentSandboxGroupContextImpl",
	"message": "No current sandbox group context to remove for thread [single-threaded-scheduled-executor-1]",
--
	"loggerName": "net.corda.sandboxgroupcontext.service.impl.SandboxGroupContextCacheImpl",
	"message": "Evicting flow sandbox for O=WholesaleBank9, L=London, C=GB [SIZE]",
--
	"loggerName": "net.corda.sandboxgroupcontext.service.impl.SandboxGroupContextCacheImpl",
	"message": "Closing flow sandbox for O=WholesaleBank1, L=London, C=GB",
--
```

Old (from ticket):
```
--
	java.lang.ClassCastException
	net.corda.ledger.utxo.flow.impl.transaction.UtxoSignedTransactionImpl cannot be cast to class net.corda.ledger.utxo.flow.impl.transaction.UtxoSignedLedgerTransaction
--
	No current sandbox group context to remove for thread [single-threaded-scheduled-executor-1]
	at net.corda.sandboxgroupcontext.service.impl.CurrentSandboxGroupContextImpl.remove(CurrentSandboxGroupContextImpl.kt:60)
--
	Failed to close SandboxGroupContext
	java.security.AccessControlException: access denied (\"org.osgi.framework.AdaptPermission\" \"org.osgi.framework.wiring.FrameworkWiring\" \"adapt\")
--
	Bundle com.r3.digital-currencies.contracts could not be uninstalled
	"java.security.AccessControlException: access denied (\"org.osgi.framework.AdminPermission\" \"(id=637)\" \"lifecycle\")
--
```

We appear to be correctly unloading the bundles now.

The text "java.security.AccessControlException: access denied" doesn't appear anywhere in the logs now, and that is exactly what I'd expect this to resolve.

**Review by rick-r3 — APPROVED** [review]

**Pull request merged**